src/environment/msfs/MsfsAddonEnvironment.cpp

```diff
@@ -0,0 +1,212 @@
+/*
+ *   AviTab - Aviator's Virtual Tablet
+ *   Copyright (C) 2018 Folke Will <folko@solhost.org>
+ *
+ *   This program is free software: you can redistribute it and/or modify
+ *   it under the terms of the GNU Affero General Public License as published by
+ *   the Free Software Foundation, either version 3 of the License, or
+ *   (at your option) any later version.
+ *
+ *   This program is distributed in the hope that it will be useful,
+ *   but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *   GNU Affero General Public License for more details.
+ *
+ *   You should have received a copy of the GNU Affero General Public License
+ *   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+#include <cmath>
+#include "MsfsAddonEnvironment.h"
+#include "src/Logger.h"
+#include "src/platform/Platform.h"
+#include "src/libxdata/world/World.h"
+#include <tchar.h>
+#include <stdio.h>
+#include <strsafe.h>
+
+#define MSFS_VERBOSE_LOGGING 0
+
+namespace avitab {
+
+MsfsAddonEnvironment::MsfsAddonEnvironment()
+:   StandAloneEnvironment(),
+    hSimConnect(NULL),
+    nextSimUpdate(0),
+    range(10)
+{
+    resetLocations();
+    setZoomLevel(12);
+}
+
+MsfsAddonEnvironment::~MsfsAddonEnvironment()
+{
+    if (hSimConnect != NULL) {
+        (void)SimConnect_Close(hSimConnect);
+    }
+}
+
+void MsfsAddonEnvironment::eventLoop()
+{
+    while (driver->handleEvents()) {
+        runEnvironmentCallbacks();
+        lastDrawTime = driver->getLastDrawTime() / 1000.0;
+
+        auto t = GetTickCount64();
+        if (t >= nextSimUpdate) {
+            if (hSimConnect == NULL) {
+                tryConnectToMsfsSim();
+                nextSimUpdate = t + 5000; // try again in 5s
+            } else {
+                retrieveMsfsObjectData();
+                nextSimUpdate = t + 1000; // next update in 1s
+            }
+        }
+    }
+    driver.reset();
+}
+
+AircraftID MsfsAddonEnvironment::getActiveAircraftCount()
+{
+    return 1 + otherLocations.size();
+}
+
+Location MsfsAddonEnvironment::getAircraftLocation(AircraftID id)
+{
+    Location loc;
+
+    if (id == 0) {
+        return userLocation;
+    } else {
+        return otherLocations[id-1];
+    }
+
+    return loc;
+}
+
+void MsfsAddonEnvironment::setZoomLevel(int zoom)
+{
+    // for zoom levels up to 8 request max range of 200km
```

Comment:
I think the meaning of the zoom level depends on the map type that's being used. For EPSG tiles, the meaning is fixed. But when using charts as a map, the meaning changes. Maybe `OverlayedMap::getMapWidthNM` is easier to deal with than the raw zoom level.

Reply:
Have discussed this with David, and agreed that knowing the current range of the map isn't really important, and better to just request aircraft in a fixed range. So we have removed the changes to the zoom level APIs.